libnetdata/dictionary/dictionary.c

```diff
@@ -253,78 +582,362 @@ int dictionary_del(DICTIONARY *dict, const char *name) {
     return ret;
 }
 
+// ----------------------------------------------------------------------------
+// traversal with loop
+
+void *dictionary_foreach_start(DICTFE *dfe, DICTIONARY *dict) {
+    if(unlikely(!dfe || !dict)) return NULL;
+
+    dfe->dict = dict;
+    dfe->started_ut = now_realtime_usec();
+
+    dictionary_read_lock(dict);
+
+    NAME_VALUE *nv = dict->first_item;
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+void *dictionary_foreach_next(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return NULL;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(likely(nv)) {
+        // TODO: decrement reference counter on nv
+
+        // get the next item
+        nv = nv->next;
+    }
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv->next;
+    }
+    else {
+        dfe->name = NULL;
+        dfe->value = NULL;
+    }
+
+    return dfe->value;
+}
+
+usec_t dictionary_foreach_done(DICTFE *dfe) {
+    if(unlikely(!dfe || !dfe->dict)) return 0;
+
+    NAME_VALUE *nv = (NAME_VALUE *)dfe->last_position_index;
+    if(nv) {
+        // TODO: decrement reference counter on nv
+        ;
```

Comment:
This comment applies here and also for another `TODO` in the previous function. We also have functions to decrement, if we are also missing calls for these functions, please, add also a call here.

Reply:
same as above